### PR TITLE
复现reid-strong-baseline

### PR DESCRIPTION
**Pull request by weisy11**

### CODE REVIEW

File: ppcls/configs/Pedestrian/strong_baseline_m1.yaml [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  save_interval: 40
+  eval_during_train: True
+  eval_interval: 10
+  epochs: 120
+  print_batch_step: 20
+  use_visualdl: False
+  warmup_by_epoch: True
+  eval_mode: "retrieval"
+  re_ranking: False
+  feat_from: "neck" # 'backbone' or 'neck'
```
Comment (RainFrost1) [inline ppcls/configs/Pedestrian/strong_baseline_m1.yaml]: 这个参数是刚加的吗

File: ppcls/engine/train/train.py [hunk 2]
```diff
@@ -63,12 +63,15 @@ def train_epoch(engine, epoch_id, print_batch_step):
...
         # clear grad
         for i in range(len(engine.optimizer)):
             engine.optimizer[i].clear_grad()
+
         # step lr
-        for i in range(len(engine.lr_sch)):
-            engine.lr_sch[i].step()
+        if engine.config["Global"].get("warmup_by_epoch", False) is False:
+            for i in range(len(engine.lr_sch)):
+                engine.lr_sch[i].step()
```
Comment (RainFrost1) [inline ppcls/engine/train/train.py:74]: 如果为True的话，lr就不更新了？

File: ppcls/loss/centerloss.py [hunk 3]
```diff
@@ -1,54 +1,74 @@
...
+
+        Args:
+            input (Dict[str, paddle.Tensor]): {'features': (batch_size, feature_dim), ...}.
+            target (paddle.Tensor): ground truth label with shape (batch_size, ).
+
+        Returns:
+            Dict[str, paddle.Tensor]: {'CenterLoss': loss}.
         """
-        feats = input["features"]
+        feats = input['backbone']
```
Comment (RainFrost1) [inline ppcls/loss/centerloss.py]: 一般其他的loss中feats都是input['features']，为了统一，建议加工参数控制一下吧，默认是‘features'

File: ppcls/loss/triplet.py [hunk 4]
```diff
@@ -24,7 +25,7 @@ def forward(self, input, target):
             inputs: feature matrix with shape (batch_size, feat_dim)
             target: ground truth labels with shape (num_classes)
         """
-        inputs = input["features"]
+        inputs = input["backbone"]
```
Comment (RainFrost1) [inline ppcls/loss/triplet.py]: 同上，加个参数控制一下，默认是’features'

File: ppcls/loss/triplet.py [hunk 5]
```diff
@@ -135,3 +136,122 @@ def forward(self, input, target):
         y = paddle.ones_like(dist_an)
         loss = self.ranking_loss(dist_an, dist_ap, y)
         return {"TripletLoss": loss}
+
+
+class TripletLossV3(nn.Layer):
```
Comment (RainFrost1) [inline ppcls/loss/triplet.py]: V3跟V2有什么区别吗？还是说V3只是加了hard ming？

File: ppcls/optimizer/__init__.py [hunk 6]
```diff
@@ -60,19 +60,20 @@ def build_optimizer(config, epochs, step_each_epoch, model_list=None):
     """NOTE:
     Currently only support optim objets below.
     1. single optimizer config.
-    2. next level uner Arch, such as Arch.backbone, Arch.neck, Arch.head.
-    3. loss which has parameters, such as CenterLoss.
+    2. model(entire Arch), backbone, neck, head.
+    3. loss(entire Loss), specific loss listed in ppcls/loss/__init__.py.
```
Comment (RainFrost1) [inline ppcls/optimizer/__init__.py]: optimizer功能是在另外一个pr里吧，不要跟这跟混了。一个pr一个功能，方便管理

File: ppcls/optimizer/learning_rate.py [hunk 7]
```diff
@@ -198,34 +215,69 @@ def __init__(self,
...
                  last_epoch=-1,
                  **kwargs):
         super().__init__()
         if warmup_epoch >= epochs:
             msg = f"When using warm up, the value of \"Global.epochs\" must be greater than value of \"Optimizer.lr.warmup_epoch\". The value of \"Optimizer.lr.warmup_epoch\" has been set to {epochs}."
             logger.warning(msg)
             warmup_epoch = epochs
-        self.boundaries = [step_each_epoch * e for e in decay_epochs]
+        self.boundaries_steps = [step_each_epoch * e for e in decay_epochs]
+        self.boundaries_epoch = decay_epochs
```
Comment (RainFrost1) [inline ppcls/optimizer/learning_rate.py:228]: warmup_by_epoch，就是为了实现前多个epoch作为warmup 的功能吗？这个功能，本来就支持的

File: ppcls/configs/Pedestrian/strong_baseline_m1.yaml (hunk 1 above)
Comment (HydrogenSulfate) [reply ppcls/configs/Pedestrian/strong_baseline_m1.yaml]: 这个参数应该是给没有neck得模型用，没有neck得话或者没写这个参数，代码是默认拿backbone的特征来用

File: ppcls/engine/train/train.py (hunk 2 above)
Comment (HydrogenSulfate) [reply ppcls/engine/train/train.py:74]: 如果为True的话，trian.py内部的lr不会更新，在engine.py的train方法内进行更新，https://github.com/PaddlePaddle/PaddleClas/pull/1819/files#diff-5ce27639f5185a27f60e70828656bbace2e48d60bccc357d29cbb7482f7da6c7R317

File: ppcls/loss/triplet.py (hunk 5 above)
Comment (HydrogenSulfate) [reply ppcls/loss/triplet.py]: 测试了一下，输出是完全一致的，我改用V2吧

File: ppcls/optimizer/learning_rate.py (hunk 7 above)
Comment (HydrogenSulfate) [reply ppcls/optimizer/learning_rate.py:228]: 由于dataloader的长度会变化，所以原有的用一个epoch的iter数计算存在一定误差，无法保证学习率变化出现在变化点的第一个iter处，而可能会出现在中间位置

File: ppcls/loss/triplet.py (hunk 4 above)
Comment (HydrogenSulfate) [reply ppcls/loss/triplet.py]: Done

File: ppcls/loss/centerloss.py (hunk 3 above)
Comment (HydrogenSulfate) [reply ppcls/loss/centerloss.py]: Done

File: ppcls/optimizer/__init__.py (hunk 6 above)
Comment (HydrogenSulfate) [reply ppcls/optimizer/__init__.py]: Done

File: ppcls/loss/centerloss.py [hunk 8]
```diff
@@ -1,54 +1,79 @@
...
+    Args:
+        num_classes (int): number of classes.
+        feat_dim (int): number of feature dimensions.
+        feat_from (str): features from backbone or neck
+    """
+
+    def __init__(self,
+                 num_classes: int,
+                 feat_dim: int,
+                 feat_from: str='backbone'):
```
Comment (RainFrost1) [inline ppcls/loss/centerloss.py]: 默认是‘features’吧

Comment (RainFrost1) [inline ppcls/loss/centerloss.py]: 默认是‘features’吧

File: ppcls/loss/triplet.py [hunk 9]
```diff
@@ -12,9 +26,13 @@ class TripletLossV2(nn.Layer):
         margin (float): margin for triplet.
     """
 
-    def __init__(self, margin=0.5, normalize_feature=True):
+    def __init__(self,
+                 margin=0.5,
+                 normalize_feature=True,
+                 feat_from='backbone'):
```
Comment (RainFrost1) [inline ppcls/loss/triplet.py]: 默认是‘features‘，metric learning相关loss输入都是features，保持统一吧

Comment (RainFrost1) [inline ppcls/loss/triplet.py]: 默认是‘features‘，metric learning相关loss输入都是features，保持统一吧

**Review by RainFrost1 — DISMISSED** [review]

**Review by RainFrost1 — APPROVED** [review]
